fil_fungible_token/src/token/mod.rs

```diff
@@ -278,8 +285,380 @@ where
-        state.save(&self.bs)?;
-
+        self.save_state(&state)?;
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod test {
+    use cid::Cid;
+    use fvm_shared::econ::TokenAmount;
+    use num_traits::Zero;
+
+    use crate::blockstore::SharedMemoryBlockstore;
+
+    use super::Token;
+
+    #[test]
+    fn it_instantiates() {
+        let state_cid = Cid::default();
+        let mut token = Token::new(SharedMemoryBlockstore::new(), state_cid);
+        token.init_state().expect("state init failed");
+        token.load_state();
+    }
+
+    #[test]
+    fn it_persists_loads() {
+        let state_cid = Cid::default();
+        let bs = SharedMemoryBlockstore::new();
+        let mut token = Token::new(bs.clone(), state_cid);
+        token.init_state().unwrap();
+
+        // simulate actor calls mint method
+        let treasury = 1;
+        token.mint(treasury, TokenAmount::from(100)).unwrap();
+        let balance = token.balance_of(treasury).unwrap();
+        assert_eq!(balance, TokenAmount::from(100));
+
+        // here the actor would save the cid to it's state tree
+        // TODO: state_cid needs to be exported so that actors can keep track of it
+        let new_state_cid = token.state_cid;
+
+        // simulate actor is invoked again: load a new instance of the Token from the blockstore
+        let token_clone = Token::new(bs, new_state_cid);
+        let balance_from_cloned = token_clone.balance_of(treasury).unwrap();
+        // it has the balances from before
+        assert_eq!(balance_from_cloned, TokenAmount::from(100));
+    }
+
+    #[test]
+    fn it_mints() {
+        let state_cid = Cid::default();
+        let mut token = Token::new(SharedMemoryBlockstore::new(), state_cid);
+        token.init_state().unwrap();
+
+        let treasury = 1;
+        token.mint(treasury, TokenAmount::from(1_000_000)).unwrap();
+
+        let balance = token.balance_of(treasury).unwrap();
+        assert_eq!(balance, TokenAmount::from(1_000_000));
+
+        let total_supply = token.total_supply();
+        assert_eq!(total_supply, TokenAmount::from(1_000_000));
+    }
+
+    #[test]
+    fn it_burns() {
+        let state_cid = Cid::default();
+        let mut token = Token::new(SharedMemoryBlockstore::new(), state_cid);
+        token.init_state().unwrap();
+
+        let treasury = 1;
+        let mint_amount = TokenAmount::from(1_000_000);
+        let burn_amount = TokenAmount::from(600_000);
+        token.mint(treasury, mint_amount).unwrap();
+        token.burn(treasury, treasury, burn_amount).unwrap();
+
+        // total supply decreased
+        let total_supply = token.total_supply();
+        assert_eq!(total_supply, TokenAmount::from(400_000));
+
+        // treasury balance decreased
+        let balance = token.balance_of(treasury).unwrap();
+        assert_eq!(balance, TokenAmount::from(400_000));
+    }
+
+    #[test]
+    fn it_allows_delegated_burns() {
```

Comment:
Nit: move complicated tests like delegation below simple ones like negatives and other edge cases.